### DISCUSSION TITLE
Create text sections with ^L (form feed)

### DISCUSSION THREAD

**Discussion by terminatorul** [vim/vim#15292] [Q&A]
Hello

I find source code (and other text) in a file can be divided into pages (sections) using form feed ^L characters, and compilers like gcc support this. The form feed character is a form of white space. I think this can be a useful feature for better organizing code in a source file, and I don't see why this feature is almost never used

Is there a way to make Vim show these markers as pages, for example with a horizontal line for the ^L characters ?

**Comment by knarrff** [comment]
Hi @terminatorul 

I don't think it is easy to show the markers on the entire line. You could use syntax highlighting to at least make the ^L stand out:

```
highlight PageHighlight ctermbg=darkgreen
call matchadd('PageHighlight', '^L *')
```

(Note: the ^L is <ctrl+l>, not "hat l")

The pattern includes possible spaces after ```^L```, so you could, if you wanted, use that to "simulate" a longer marker by inserting spaces after each ^L, but this is optional, as it might be a little strange in the source. I did not find a way to highlight text that isn't there: in order to have a whole line highlighted, the line must have that much text (thus, the spaces), as highlighting affects text, not lines.

**Comment by terminatorul** [comment]
Thank you,

I found an answer on stackoverflow that uses folds, and I applied that if the source file does not have a `'foldmethod'` already (the intent is to apply this if there are no folds at all). The visual is nice, a true horizontal ruler in the source code, based on Unicode character `"\u2015"`.

But I was hoping for a native solution in Vim, like a true Vim option I could enable to make pages visible.

```vim
function! FoldPageFeed()
    setlocal foldmethod=expr
    setlocal foldexpr=getline(v:lnum)[0]==\"\\<c-l>\"
    setlocal foldminlines=0
    let &l:foldtext = '"\u2015"'
    setlocal foldlevel=0
    set foldclose=all
    eval "set fillchars+=fold:\u2015"->execute()

    nnoremap <buffer> <C-S-F> :eval "\f"->search('zWe')<CR>
    nnoremap <buffer> <C-S-B> :eval "\f"->search('bzWe')<CR>
endfunction

nmap ,ff :call FoldPageFeed()<CR>

let s:autocmd = { }

let s:autocmd['group'] = 'FoldFormFeed'
let s:autocmd['event'] = 'FileType'
let s:autocmd['pattern'] = [ 'c', 'cpp', 'perl', 'make' ]
let s:autocmd['cmd'] = 'if &l:foldmethod == ''manual'' | call FoldPageFeed() | endif'

eval [ s:autocmd ]->autocmd_add()

```
(see this gist https://gist.github.com/terminatorul/06f5e8ec5b291127a3d77798a12d0d21 for using one-line comments as section titles)

The feature is very good if your codebase makes use of ^L for paging. I did not have a chance to add form feed characters myself in a new source file written by me.

![image](https://github.com/user-attachments/assets/e7de24a6-9ee9-415c-ae6e-a97560d25a4b)
